src/Language/DifferentialDatalog/Validate.hs

```diff
@@ -557,3 +523,46 @@ exprCheckMatchPatterns d ctx e@(EMatch _ x cs) = do
     assert (consTreeEmpty ct) (pos x) "Non-exhaustive match patterns"
 
 exprCheckMatchPatterns _ _   _               = return ()
+
+-- Automatically insert string conversion functions in the Concat
+-- operator:  '"x:" ++ x', where 'x' is of type int becomes 
+-- '"x:" ++ int_2string(x)'.
+progInjectStringConversions :: (MonadError String me) => DatalogProgram -> me DatalogProgram
+progInjectStringConversions d = progExprMapCtxM d (exprInjectStringConversions d)
+
+exprInjectStringConversions :: (MonadError String me) => DatalogProgram -> ECtx -> ENode -> me Expr
+exprInjectStringConversions d ctx e@(EBinOp p Concat l r) | (te == tString) && (tr /= tString) = do
+    -- find string conversion function
+    fname <- case tr of
+                  TBool{}     -> return bUILTIN_2STRING_FUNC
+                  TInt{}      -> return bUILTIN_2STRING_FUNC
+                  TString{}   -> return bUILTIN_2STRING_FUNC
+                  TBit{}      -> return bUILTIN_2STRING_FUNC
+                  TUser{..}   -> return $ mk2string_func typeName
+                  TOpaque{..} -> return $ mk2string_func typeName
+                  TTuple{}    -> err (pos r) "Automatic string conversion for tuples is not supported"
+                  TVar{..}    -> err (pos r) $ 
```

Comment:
Is this a type variable?
If so maybe the conversion should be done after unification, and there shouldn't be any more type variables left.

Reply:
This is doable, but not easy. Problem is, we will generate polymorphic Rust functions, with unification (specialization, actually) being done by the Rust compiler. In a richer type system, we'd be able to write that the type variable must be of a printable type, but I really don't want to add type classes or inheritance to the language. Alternatively, we could infer that a type variable must be printable from function implementation, and convert its 2string function to the Rust Display trait implementation.

I'd rather not spend time on this now, as I want to make progress. How about I add this as an issue for now?